core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java

```diff
@@ -0,0 +1,433 @@
+        return !completionFuture.isDone();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        if (isOpen()) {
+            return false;
+        }
+        final SurroundingSubscriber<T> surroundingSubscriber = this.surroundingSubscriber;
+        return surroundingSubscriber == null || !surroundingSubscriber.publishedAny;
+    }
+
+    @Override
+    public long demand() {
+        final SurroundingSubscriber<T> surroundingSubscriber = this.surroundingSubscriber;
+        if (surroundingSubscriber != null) {
+            return surroundingSubscriber.requested;
+        } else {
+            return 0;
+        }
+    }
+
+    @Override
+    public CompletableFuture<Void> whenComplete() {
+        return completionFuture;
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super T> subscriber, EventExecutor executor,
+                          SubscriptionOption... options) {
+        requireNonNull(subscriber, "subscriber");
+        requireNonNull(executor, "executor");
+        requireNonNull(options, "options");
+
+        if (!subscribedUpdater.compareAndSet(this, 0, 1)) {
+            subscriber.onSubscribe(NoopSubscription.get());
+            subscriber.onError(new IllegalStateException("Only single subscriber is allowed!"));
+            return;
+        }
+
+        if (executor.inEventLoop()) {
+            subscribe0(subscriber, executor, options);
+        } else {
+            executor.execute(() -> subscribe0(subscriber, executor, options));
+        }
+    }
+
+    private void subscribe0(Subscriber<? super T> subscriber, EventExecutor executor,
+                            SubscriptionOption... options) {
+
+        final SurroundingSubscriber<T> surroundingSubscriber = new SurroundingSubscriber<>(
+                head, publisher, tail, subscriber, executor, completionFuture, options);
+        this.surroundingSubscriber = surroundingSubscriber;
+        subscriber.onSubscribe(surroundingSubscriber);
+
+        // To make sure to close the SurroundingSubscriber when this is aborted.
+        if (completionFuture.isCompletedExceptionally()) {
+            completionFuture.exceptionally(cause -> {
+                surroundingSubscriber.close(cause);
+                return null;
+            });
+        }
+    }
+
+    @Override
+    public void abort() {
+        abort(AbortedStreamException.get());
+    }
+
+    @Override
+    public void abort(Throwable cause) {
+        requireNonNull(cause, "cause");
+
+        // `completionFuture` should be set before `SurroundingSubscriber` publishes data
+        // to guarantee the visibility of the abortion `cause` after
+        // SurroundingSubscriber is set in `subscriber0()`.
+        completionFuture.completeExceptionally(cause);
+
+        final SurroundingSubscriber<T> surroundingSubscriber = this.surroundingSubscriber;
+        if (surroundingSubscriber != null) {
+            surroundingSubscriber.close(cause);
+        }
```

Comment:
We should abort the upstream if `surroundingSubscriber` is null in case this publisher is not subscribed.
```suggestion

        if (subscribedUpdater.compareAndSet(this, 0, 1)) {
            upstream.abort(cause);
            if (head != null) {
                StreamMessageUtil.closeOrAbort(head, cause);
            }
            if (tail != null) {
                StreamMessageUtil.closeOrAbort(tail, cause);
            }
            return;
        }

        completionFuture.completeExceptionally(cause);
        final SurroundingSubscriber<T> surroundingSubscriber = this.surroundingSubscriber;
        if (surroundingSubscriber != null) {
            surroundingSubscriber.close(cause);
        }
```

Reply:
✅ updated! nice suggestion 👍